### PR TITLE
[RemoteMirror] Add a call to get the demangled name for a typeref

### PR DESCRIPTION
**Pull request by mikeash**
rdar://problem/59909982

### CODE REVIEW

**Review by dcci — COMMENTED** [review]
Adrian worked on something very close to this so maybe he wants to check this one out.

**Comment by adrian-prantl** [comment]
> Adrian worked on something very close to this so maybe he wants to check this one out.

This is the merged PR that Davide is referring to:
https://github.com/apple/swift/pull/29139

**Comment by adrian-prantl** [comment]
... which is incidentally the getDemangling() API being used here :-)

**Comment by mikeash** [comment]
Yeah, I happened across your PR and said, hey, this is just what the doctor ordered!

**Comment by mikeash** [comment]
@swift-ci please test and merge

**Comment by mikeash** [comment]
@swift-ci please test

**Comment by mikeash** [comment]
@adrian-prantl Please take a look at the changes to `stdlib/public/Reflection/TypeRef.cpp ` that I just pushed.

File: include/swift/SwiftRemoteMirror/SwiftRemoteMirrorLegacyInterop.h [hunk 1]
```diff
@@ -354,7 +362,7 @@ swift_reflection_interop_libraryOwnsAddress(
   // Search the images list to see if the address is in one of them.
   struct SwiftReflectionInteropContextLegacyImageRangeList *Node =
     ContextRef->LegacyImageRangeList;
-  while (Node != nullptr) {
+  while (Node != NULL) {
```
Comment (adrian-prantl) [inline include/swift/SwiftRemoteMirror/SwiftRemoteMirrorLegacyInterop.h]: Is this a C-compatible header?

Comment (adrian-prantl) [reply include/swift/SwiftRemoteMirror/SwiftRemoteMirrorLegacyInterop.h]: Also, why not `!Node`?

File: stdlib/public/Reflection/TypeRef.cpp [hunk 2]
```diff
@@ -383,6 +383,16 @@ class DemanglingForTypeRef
...
   DemanglingForTypeRef(Demangle::Demangler &Dem) : Dem(Dem) {}
 
+  Demangle::NodePointer visit(const TypeRef *typeRef) {
+    auto node = TypeRefVisitor<DemanglingForTypeRef,
+                                Demangle::NodePointer>::visit(typeRef);
+
+    // Wrap all nodes in a Type node, as consumers generally expect.
+    auto typeNode = Dem.createNode(Node::Kind::Type);
+    typeNode->addChild(node, Dem);
+    return typeNode;
```
Comment (adrian-prantl) [inline stdlib/public/Reflection/TypeRef.cpp]: I think this makes sense, thanks!

File: include/swift/SwiftRemoteMirror/SwiftRemoteMirrorLegacyInterop.h (hunk 1 above)
Comment (mikeash) [reply include/swift/SwiftRemoteMirror/SwiftRemoteMirrorLegacyInterop.h]: It is, analogous to `SwiftRemoteMirror.h`. As for why not `!Node`, you'd have to ask my brain, and it hasn't been taking questions lately.

**Comment by mikeash** [comment]
@swift-ci please test

**Comment by mikeash** [comment]
@swift-ci please test and merge

**Comment by mikeash** [comment]
@swift-ci please test and merge

**Pull request merged**